spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,651 @@
+
+## Technical Specification
+
+### General Design
+
+<img src="./ibcswap.png"/>
+
+A maker offers token A in exchange for token B by making an order. The order specifies the quantity and price of exchange, and sends the offered token A to the maker chain's escrow account. Any taker on a different chain with token B can accept the offer by taking the order. The taker sends the desired amount of token B to the taker chain's escrow account. The escrow account on each respective chain transfers the corresponding token amounts to each user's receiving address, without requiring the usual IBC transfer.
+
+An order without takers can be cancelled.  This enables users to rectify mistakes, such as inputting an incorrect price or taker address.  Upon cancellation escrowed tokens will be refunded. 
+
+When making or taking an order, a timeout window is specified in the relayed data packet.  A timeout will result in escrowed tokens refunded back.  This timeout window is customizable.
+
+### Data Structures
+
+Only one packet data type is required: `AtomicSwapPacketData`, which specifies the swap message type, data (protobuf marshalled) and a memo field.
+
+```typescript
+enum SwapMessageType {
+  // Default zero value enumeration
+  TYPE_UNSPECIFIED = 0,
+  TYPE_MSG_MAKE_SWAP = 1,
+  TYPE_MSG_TAKE_SWAP = 2,
+  TYPE_MSG_CANCEL_SWAP = 3,
+}
+```
+
+```typescript
+// AtomicSwapPacketData is comprised of a swap message type, raw transaction and optional memo field.
+interface AtomicSwapPacketData {
+  type: SwapMessageType
+  data: bytes
+  memo: string
+}
+```
+```typescript
+type AtomicSwapPacketAcknowledgement = AtomicSwapPacketSuccess | AtomicSwapPacketError
+interface AtomicSwapPacketSuccess {
+  // This is binary 0x01 base64 encoded
+  result: "AQ=="
+}
+
+interface AtomicSwapPacketError {
+  error: string
+}
+```
+
+All `AtomicSwapPacketData` will be forwarded to the corresponding message handler to execute according to its type. There are 3 types:
+
+```typescript
+interface MakeSwapMsg {
+  // the port on which the packet will be sent, specified by the maker when the message is created
+  sourcePort: string
+  // the channel on which the packet will be sent, specified by the maker when the message is created
+  sourceChannel: string
+  // the tokens to be exchanged
+  sellToken: Coin
+  buyToken: Coin
+  // the maker's address
+  makerAddress: string
+  // the maker's address on the taker chain
+  makerReceivingAddress string
+  // if desiredTaker is specified,
+  // only the desiredTaker is allowed to take this order
+  // this is the address on the taker chain
+  desiredTaker: string
+  creationTimestamp: uint64
+  expirationTimestamp: uint64
+  timeoutHeight: Height
+  timeoutTimestamp: uint64
+  nonce: string // a random 6 digits string
+}
+```
+
+```typescript
+interface TakeSwapMsg {
+  // the channel on which the packet will be sent, specified by the taker when the message is created
```

Comment:
In order to ensure that the taker message is sent to the maker chain that created the exchange, we save the `destinationPort` and `destinationChannel` of `makeMsg` in the Order (should be the `sourcePort` and `sourceChannel` of takeMsg). Here we originally wanted to use the `sourceChannel` as part of the storage key to avoid collisions, since we have `nonce` or `sequence`,  I'm going to remove `sourceChannel` from both takeMsg and CancelMsg.